src/borrow.rs

```diff
@@ -0,0 +1,579 @@
+//! Types to safely create references into the interior of the NumPy arrays
```

Comment:
This document is well written and I enjoyed reading it. However, I fear this can be a bit overwhelming for beginners. So, I propose to re-organize the document in a way like:
```markdown
## Summary
The position we take, and the resulting design.
## Some background/reasoning about the design
The reason why this design is useful/reasonable is considering the Rust/Python boundary.
## Current limitation
```

Reply:
I pushed an updated version where I tried to restructure the module-level documentation along the proposed lines.